### ISSUE TITLE
Address for accepting donations to a Project or Drip List

### ISSUE THREAD

**Issue by brandonhaslegs** [drips-network/app#824]
Discussion is [here](https://discord.com/channels/1096003917717983252/1176167908859191387)

Essentially the idea is generating an address that users can send tokens to which automatically sends to the Drip List or Project. The benefit of having this is users don't have to make an account (connect a wallet, find the project/list, top up funds, etc) to participate in Drips.

**Comment by efstajas** [comment]
Quick rundown of the technical idea as I understand it…

We'd need:

1) a proxy contract that allows the owner of a particular drips account ID to access all ERC-20 funds it holds, and call Drips functions on its Address Driver account ID. **Ideally, somehow anyone could call the functions required for pushing funds from the contract's balances to the respective Drip List or Project's**, so that funds donated to the proxy contract aren't stuck until the owner comes along to collect them. @CodeSandwich can we do that? 🤔 
2) a factory that can deploy those proxy contracts at a deterministic address derived from the respective Drips account ID it's being deployed for

With this in place, we can add to the app:

- A call that deploys such a proxy contract whenever a new Drip List is created or Project claimed
- The ability to see the "donation address" for drip lists and projects, which can accept one-time donations. We can do this even for unclaimed projects.
- In case the particular Project or Drip List has been created before the proxy contract creation call was added, a way to retroactively deploy it and push any funds held by the contract at time of creation to the respective Drip List / Project

**Comment by CodeSandwich** [comment]
> a proxy contract that allows the owner of a particular drips account ID to access all ERC-20 funds it holds, and call Drips functions on its Address Driver account ID.

I think that the "access all ERC-20" and "call Drips functions" is not impossible in a generic solution, because there's no clear definition of the "account ownership" in Drips outside of each driver code. What we **can** do in a generic way, is to only allow `give`ing to the account ID for which the proxy contract has been created.

The proxy contract derived from `accountId` will only expose a single function: `function give(erc20)`. It will query `erc20` for the `amt` it holds, and then it will call `AddressDriver.give(accountId, erc20, amt)`, that's it. This approach is simple, generic and permissionless, there's no harm in strangers calling `give`. There's also no account ID ownership resolution, because there are no functionalities that would require any authorization.

The downside is that in order to track down who exactly made the donation, we'll need to check if the address that called `AddressDriver.give(...)` is a proxy (easy if we use a factory to deploy the proxies), and if it is, scan the `Transferred` events for the ERC-20 we're interested in.

> A call that deploys such a proxy contract whenever a new Drip List is created or Project claimed

It could be easier to do the deployment on the first usage, because it won't add an extra step in the project claiming or list creation flow. The deployment on the first usage must be implemented anyway, because of point 3, `a way to retroactively deploy `.

**Comment by efstajas** [comment]
> The downside is that in order to track down who exactly made the donation, we'll need to check if the address that called AddressDriver.give(...) is a proxy (easy if we use a factory to deploy the proxies), and if it is, scan the Transferred events for the ERC-20 we're interested in.

Hm, I haven't thought about that. I suppose in order to display balances for these proxy contracts we're going to need to make use of some ERC-20 balances API. Our backend obviously can't track all transfer events emitted by all ERC-20 contracts…

> It could be easier to do the deployment on the first usage, because it won't add an extra step in the project claiming or list creation flow.

Just to be clear, I understood that the goal is that each individual Drip List and Project gets its own separate proxy contract and thus donation address, right? So we could just batch the contract creation in with the existing batches for claiming a project and creating a Drip List.

**Comment by CodeSandwich** [comment]
> the goal is that each individual Drip List and Project gets its own separate proxy contract and thus donation address, right?

Yes, but we don't need to deploy anything until we want to call `give(erc20)`. We can calculate the deterministic proxy address off-chain, ask people to send funds there. If somebody will look up the address, they will see an address with no code, as if it was an EOA, a perfectly valid recipient of ERC-20 and plain ETH. Only when somebody wants to call `give` for the fist time, we need to deploy the contract. It will get the same address as we calculated off-chain, and at the moment of the deployment it will already hold all the assets, including plain ETH.

**Comment by efstajas** [comment]
We have a solution for this on-chain now: Giver contracts.

Let's use this task to track surfacing these addresses for Drip Lists & Projects. First of all, we need design for how the donation addresses appear on Projects & Drip Lists.

In theory, not much should be required for handling displaying of donations via Giver in the app, because they're just standard Gives that we already show as One-Time Donations. However, as-is, the source of donations would show as the Giver contract itself, rather than the original donator. So, we need at least some additional logic on API / Event Processor that:
- ignores Gives from Giver addresses for creation of One-Time Donation support items, and
- creates One-Time Donation support items for incoming ERC-20 transfers to any Giver address
... we might want to do this at runtime on API level.

**Comment by brandonhaslegs** [comment]
So just thinking of showing an extra button here 
<img width="728" alt="Image" src="https://github.com/user-attachments/assets/df563790-537e-4c2a-91a8-e22d8f884a9f" />

which would then open a popup that displays the address for copying
![Image](https://github.com/user-attachments/assets/8c0fc789-7771-4c6c-9827-606bbbcb25db)

**Comment by efstajas** [comment]
@brandonhaslegs i think it'd be much nicer to just display the address & copy button right there. bc why not? like this it's hidden behind two clicks, also you could only see it after connecting your wallet

**Comment by efstajas** [comment]
also fyi — the funds would be moved to the project on the next settlement date, so the same thing as with OTDs applies. There should probably be a "Learn more" or ( i ) thingie next to the addresses to explain this

**Comment by brandonhaslegs** [comment]
Okay but should it also stay in that dropdown?

Here's what it might look like to display it under the support card

<img width="1152" alt="Image" src="https://github.com/user-attachments/assets/c86b44bd-3f00-4564-9c0a-631b799d45cd" />

And I also thought of just displaying an address as meta data, and showing a tooltip explaing you can donate to it

<img width="570" alt="Image" src="https://github.com/user-attachments/assets/f7dfb574-f044-4f99-9913-b34ecf0905b0" />

**Comment by brandonhaslegs** [comment]
This feels good 

![Image](https://github.com/user-attachments/assets/2903d6ca-bd1f-49d5-b638-8a685d47f404)
